projects/lesson-5/runtime/src/kitties.rs

```diff
@@ -0,0 +1,435 @@
+use support::{decl_module, decl_storage, ensure, StorageValue, StorageMap, dispatch::Result, Parameter, traits::Currency};
+use sr_primitives::traits::{SimpleArithmetic, Bounded, Member, Zero};
+use codec::{Encode, Decode};
+use runtime_io::blake2_128;
+use system::ensure_signed;
+use rstd::result;
+
+pub trait Trait: balances::Trait {
+	type KittyIndex: Parameter + Member + SimpleArithmetic + Bounded + Default + Copy;
+}
+
+#[derive(Encode, Decode)]
+pub struct Kitty<T: Trait> {
+	pub dna : [u8; 16],
+	pub price : T::Balance, 
+}
+
+#[cfg_attr(feature = "std", derive(Debug, PartialEq, Eq))]
+#[derive(Encode, Decode)]
+pub struct KittyLinkedItem<T: Trait> {
+	pub prev: Option<T::KittyIndex>,
+	pub next: Option<T::KittyIndex>,
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Kitties {
+		/// Stores all the kitties, key is the kitty id / index
+		pub Kitties get(kitty): map T::KittyIndex => Option<Kitty<T>>;
+		/// Stores the total number of kitties. i.e. the next kitty index
+		pub KittiesCount get(kitties_count): T::KittyIndex;
+
+		pub OwnedKitties get(owned_kitties): map (T::AccountId, Option<T::KittyIndex>) => Option<KittyLinkedItem<T>>;
+	  KittyOwner get(owner_of): map T::KittyIndex => Option<T::AccountId>;
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		/// Create a new kitty
+		pub fn create(origin) {
+			let sender = ensure_signed(origin)?;
+			let kitty_id = Self::next_kitty_id()?;
+
+			// Generate a random 128bit value
+			let dna = Self::random_value(&sender);
+
+			// Create and store kitty
+			let kitty =  Kitty {
+                dna: dna,
+                price: 0.into(),
+            };
+			Self::insert_kitty(&sender, kitty_id, kitty);
+		}
+
+		/// Breed kitties
+		pub fn breed(origin, kitty_id_1: T::KittyIndex, kitty_id_2: T::KittyIndex) {
+			let sender = ensure_signed(origin)?;
+
+			Self::do_breed(&sender, kitty_id_1, kitty_id_2)?;
+		}
+
+		// 作业：实现 transfer(origin, to: T::AccountId, kitty_id: T::KittyIndex)
+		// 使用 ensure! 来保证只有主人才有权限调用 transfer
+		// 使用 OwnedKitties::append 和 OwnedKitties::remove 来修改小猫的主人
+		pub fn transfer(origin, to: T::AccountId, kitty_id: T::KittyIndex) {
+			let sender = ensure_signed(origin)?;
+
+			let owner = Self::owner_of(kitty_id).ok_or("No owner for this kitty")?;
+        ensure!(owner == sender, "You do not own this kitty");			
+
+			Self::transfer_from(sender, to, kitty_id);
+		}
+
+		fn set_price(origin, kitty_id: T::KittyIndex, new_price: T::Balance) -> Result {
+					let sender = ensure_signed(origin)?;
+
+					ensure!(<Kitties<T>>::exists(kitty_id), "This cat does not exist");
+
+					let owner = Self::owner_of(kitty_id).ok_or("No owner for this kitty")?;
+					ensure!(owner == sender, "You do not own this cat");
```

Comment:
There is a few places have the same check, we can extract a function to do the check, 
```
ensure!(<Kitties<T>>::exists(kitty_id), "This cat does not exist");
let owner = Self::owner_of(kitty_id).ok_or("No owner for this kitty")?;
ensure!(owner == sender, "You do not own this cat");
```